akka-remote/src/main/scala/akka/remote/Remoting.scala

```diff
@@ -533,25 +533,30 @@ private[remote] class EndpointManager(conf: Config, log: LoggingAdapter) extends
       }
       Future.fold(allStatuses)(true)(_ && _) map ManagementCommandAck pipeTo sender()
 
-    case Quarantine(address, uidOption) ⇒
+    case Quarantine(address, uidToQuarantineOption) ⇒
       // Stop writers
-      endpoints.writableEndpointWithPolicyFor(address) match {
-        case Some(Pass(endpoint, _, _)) ⇒
+      (endpoints.writableEndpointWithPolicyFor(address), uidToQuarantineOption) match {
+        case (Some(Pass(endpoint, _, _)), None) ⇒
           context.stop(endpoint)
-          if (uidOption.isEmpty) {
-            log.warning("Association to [{}] with unknown UID is reported as quarantined, but " +
-              "address cannot be quarantined without knowing the UID, gating instead for {} ms.",
-              address, settings.RetryGateClosedFor.toMillis)
-            endpoints.markAsFailed(endpoint, Deadline.now + settings.RetryGateClosedFor)
-          }
-        case _ ⇒ // nothing to stop
+          log.warning("Association to [{}] with unknown UID is reported as quarantined, but " +
+            "address cannot be quarantined without knowing the UID, gating instead for {} ms.",
+            address, settings.RetryGateClosedFor.toMillis)
+          endpoints.markAsFailed(endpoint, Deadline.now + settings.RetryGateClosedFor)
+        case (Some(Pass(endpoint, Some(currentUid), _)), Some(quarantineUid)) if currentUid == quarantineUid ⇒
+          context.stop(endpoint)
+        case _ ⇒
+        // Do nothing, because either:
+        // A: we don't know yet the UID of the writer, it will be checked against current quarantine state later
+        // B: we know the UID, but it does not match with the UID to be quarantined
       }
+
       // Stop inbound read-only associations
-      endpoints.readOnlyEndpointFor(address) match {
-        case Some(endpoint) ⇒ context.stop(endpoint)
-        case _              ⇒ // nothing to stop
+      (endpoints.readOnlyEndpointFor(address), uidToQuarantineOption) match {
+        case (Some((endpoint, _)), None) ⇒ context.stop(endpoint)
+        case (Some((endpoint, currentUid)), Some(quarantineUid)) if currentUid != quarantineUid ⇒ context.stop(endpoint)
```

Comment:
you are right, there should be an equality check just like in the other case. The test does not exercise this case, because it is related to other changes.